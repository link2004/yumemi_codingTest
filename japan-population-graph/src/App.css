header {
  background-color: #282c34;
  color: white;
  overflow: hidden;
  text-align: center;
  & h1 {
    font-size: 1.5rem;
    margin: 0;
    padding: 20px 0;
  }
}

main {
  max-width: 1000px;
  margin: 0 auto;
  padding: 20px;
}

.margin-bottom-20 {
  margin-bottom: 20px;
}
