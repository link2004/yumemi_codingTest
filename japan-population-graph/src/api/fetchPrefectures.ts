const BASE_URL = 'https://opendata.resas-portal.go.jp';
const API_KEY = process.env.REACT_APP_RESAS_API_KEY;

if (API_KEY === undefined) {
  throw new Error('RESAS_API_KEYが設定されていません。');
}

export interface Prefecture {
  prefCode: number;
  prefName: string;
}

interface PrefecturesResponse {
  statusCode: string;
  message: string;
  result: Prefecture[];
}

// 都道府県データを取得する関数
export const fetchPrefectures = async (): Promise<PrefecturesResponse> => {
  try {
    const response = await fetch(`${BASE_URL}/api/v1/prefectures`, {
      headers: {
        'X-API-KEY': API_KEY,
        // 'X-API-KEY': '',
      },
    });
    const data = (await response.json()) as PrefecturesResponse;
    // data.statusCode = '400';
    if (data.statusCode === '403') {
      throw new Error('APIキーが無効です。');
    } else if (data.statusCode >= '400') {
      throw new Error(`${data.statusCode}: ${data.message}`);
    }
    return data;
  } catch (error) {
    throw new Error(`データの取得に失敗しました。${String(error)}`);
  }
};

// 都道府県コードから都道府県の名前を取得する関数
export const getPrefectureName = (prefCode: number, prefectures: Prefecture[]): string => {
  const prefecture = prefectures.find((prefecture) => prefecture.prefCode === prefCode);
  if (prefecture === undefined) {
    throw new Error('都道府県が見つかりませんでした。');
  }
  return prefecture.prefName;
};
