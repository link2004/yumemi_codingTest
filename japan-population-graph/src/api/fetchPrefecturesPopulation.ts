const BASE_URL = 'https://opendata.resas-portal.go.jp';
const API_KEY = process.env.REACT_APP_RESAS_API_KEY;

if (API_KEY === undefined) {
  throw new Error('RESAS_API_KEYが設定されていません。');
}

interface PopulationDataPoint {
  year: number;
  value: number;
}

interface PopulationDataSet {
  label: string;
  data: PopulationDataPoint[];
}

export interface PopulationResponse {
  message: null | string;
  result: {
    boundaryYear: number;
    data: PopulationDataSet[];
  };
}

export interface PrefecturePopulation {
  prefCode: number;
  populationResponse: PopulationResponse;
}

// 都道府県の人口構成を取得する
export const fetchPrefecturePopulation = async (prefCode: number): Promise<PrefecturePopulation> => {
  try {
    const response = await fetch(`${BASE_URL}/api/v1/population/composition/perYear?cityCode=-&prefCode=${prefCode}`, {
      headers: {
        'X-API-KEY': API_KEY,
      },
    });

    if (!response.ok) {
      throw new Error(`HTTP status ${response.status}`);
    }

    const populationResponse = (await response.json()) as PopulationResponse;
    const data: PrefecturePopulation = {
      prefCode,
      populationResponse,
    };
    return data;
  } catch (error) {
    throw new Error(`データの取得に失敗しました。${String(error)}`);
  }
};
