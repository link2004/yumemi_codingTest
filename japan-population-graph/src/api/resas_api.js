// const BASE_URL = "https://opendata.resas-portal.go.jp"
