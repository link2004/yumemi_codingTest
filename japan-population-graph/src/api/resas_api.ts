const BASE_URL = 'https://opendata.resas-portal.go.jp';
const API_KEY = process.env.RESAS_API_KEY;
if (API_KEY === undefined) {
  throw new Error('RESAS_API_KEYが設定されていません。');
}

interface Prefecture {
  prefCode: number;
  prefName: string;
}

interface PrefecturesResponse {
  message: string;
  result: Prefecture[];
}

// 都道府県データを取得する関数
export const fetchPrefectures = async (): Promise<PrefecturesResponse> => {
  try {
    const response = await fetch(`${BASE_URL}/api/v1/prefectures`, {
      headers: {
        'X-API-KEY': API_KEY,
      },
    });
    const data = (await response.json()) as PrefecturesResponse;
    return data;
  } catch (error) {
    throw new Error(`データの取得に失敗しました。${String(error)}`);
  }
};
