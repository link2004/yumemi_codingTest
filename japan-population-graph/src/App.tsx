import './App.css';
import React, { useEffect, useState } from 'react';
import { fetchPrefectures, type Prefecture } from './api/fetchPrefectures';
import { fetchPrefecturePopulations, type PrefecturePopulation } from './api/fetchPrefecturesPopulation';
import PrefectureCheckBoxList from './parts/prefectureCheckBoxList';
import PopulationGraph from './parts/populationGraph';
function App(): JSX.Element {
  const [prefectures, setPrefectures] = useState<Prefecture[]>([]);
  const [selectedPrefectures, setSelectedPrefectures] = useState<Prefecture[]>([]);
  const [prefecturePopulations, setPrefecturePopulations] = useState<PrefecturePopulation[]>();

  useEffect(() => {
    const fetchData = async (): Promise<void> => {
      const prefectures = await fetchPrefectures();
      setPrefectures(prefectures.result);
    };
    void fetchData();
  }, []);

  useEffect(() => {
    const fetchData = async (): Promise<void> => {
      if (selectedPrefectures.length === 0) {
        return;
      }
      const selectedPrefecturesCodes = selectedPrefectures.map((p) => p.prefCode);
      const prefecturePopulations = await fetchPrefecturePopulations(selectedPrefecturesCodes);
      setPrefecturePopulations(prefecturePopulations);
    };
    void fetchData();
  }, [selectedPrefectures]);
  return (
    <div>
      <p>React App</p>
      <span>selected Prefectures:</span>
      {selectedPrefectures.map((p) => (
        <span key={p.prefCode}>{p.prefName},</span>
      ))}
      {prefecturePopulations != null && (
        <PopulationGraph prefecturePopulations={prefecturePopulations} prefectures={prefectures} />
      )}
      <PrefectureCheckBoxList
        prefectures={prefectures}
        selectedPrefectures={selectedPrefectures}
        setSelectedPrefectures={setSelectedPrefectures}
      />
    </div>
  );
}

export default App;
