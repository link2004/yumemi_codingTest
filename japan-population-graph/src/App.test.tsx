import { render, screen } from '@testing-library/react';
import { fetchPrefectures } from './api/fetchPrefectures';
import { fetchPrefecturePopulation, type PrefecturePopulation } from './api/fetchPrefecturesPopulation';
import { fetchPrefecturePopulations } from './api/fetchPrefecturesPopulation';
import PopulationGraph from './parts/populationGraph';
import App from './App';
test('fetchPrefectures', async () => {
  await expect(fetchPrefectures()).resolves.toBeDefined();
});

test('fetchPrefecturePopulation', async () => {
  const result = await fetchPrefecturePopulation(1);
  // resultがundefinedでないことを確認
  expect(result).toBeDefined();
  expect(result.prefCode).toBe(1);
  expect(result.populationResponse).toBeDefined();
  expect(result.populationResponse.result).toBeDefined();
  expect(result.populationResponse.result.boundaryYear).toBeDefined();
  expect(result.populationResponse.result.data).toBeDefined();
  expect(result.populationResponse.result.data[0].label).toBeDefined();
  expect(result.populationResponse.result.data[0].data).toBeDefined();
  expect(result.populationResponse.result.data[0].data[0].year).toBeDefined();
  expect(result.populationResponse.result.data[0].data[0].value).toBeDefined();
});

test('fetchPrefecturePopulations', async () => {
  const result = await fetchPrefecturePopulations([1, 2]);
  expect(result).toBeDefined();
  expect(result.length).toBe(2);
  expect(result[0].prefCode).toBe(1);
  expect(result[1].prefCode).toBe(2);
  expect(result[0].populationResponse).toBeDefined();
  expect(result[1].populationResponse).toBeDefined();
});

test('Appが表示されるか', () => {
  render(<App />);
  const linkElement = screen.getByText(/React App/i);
  expect(linkElement).toBeInTheDocument();
});

test('PopulationGraphが表示されるか', () => {
  const mockPrefectures = [
    { prefCode: 1, prefName: '北海道' },
    { prefCode: 2, prefName: '青森県' },
  ];
  const mockPopulations: PrefecturePopulation[] = [
    {
      prefCode: 1,
      populationResponse: {
        message: '',
        result: {
          boundaryYear: 2020,
          data: [
            {
              label: '総人口',
              data: [
                { year: 1980, value: 12817 },
                { year: 1985, value: 12707 },
                { year: 1990, value: 12571 },
              ],
            },
            { label: '年少人口', data: [] },
          ],
        },
      },
    },
    {
      prefCode: 2,
      populationResponse: {
        message: '',
        result: {
          boundaryYear: 2020,
          data: [
            {
              label: '総人口',
              data: [
                { year: 1980, value: 12817 },
                { year: 1985, value: 12707 },
                { year: 1990, value: 12571 },
              ],
            },
            { label: '年少人口', data: [] },
          ],
        },
      },
    },
  ];

  render(<PopulationGraph prefecturePopulations={mockPopulations} prefectures={mockPrefectures} />);

  // HighchartsReactが表示されているか
  const highchartsElement = screen.getByTestId('highcharts-container');
  expect(highchartsElement).toBeInTheDocument();
});
