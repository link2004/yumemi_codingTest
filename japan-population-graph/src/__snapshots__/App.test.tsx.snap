// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot App 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <header>
      <h1>
        都道府県別の総人口推移グラフ
      </h1>
    </header>
    <main>
      <div
        class="margin-bottom-20"
      >
        <div
          class="loading"
          data-testid="population-loading"
        />
      </div>
      <div
        class="loading"
        data-testid="prefecture-loading"
      />
    </main>
  </div>
</DocumentFragment>
`;

exports[`render: App snapshot App 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <header>
      <h1>
        都道府県別の総人口推移グラフ
      </h1>
    </header>
    <main>
      <div
        class="margin-bottom-20"
      >
        <div
          class="loading"
          data-testid="population-loading"
        />
      </div>
      <div
        class="loading"
        data-testid="prefecture-loading"
      />
    </main>
  </div>
</DocumentFragment>
`;
